app/src/main/java/com/android/bookswap/model/BooksFirestoreRepository.kt

```diff
@@ -0,0 +1,101 @@
+  // If the user is authenticated, it triggers the OnSuccess callback
+  override fun init(OnSucess: () -> Unit) {
+    Firebase.auth.addAuthStateListener {
+      if (it.currentUser != null) {
+        OnSucess()
+      }
+    }
+  }
+  // Generates and returns a new unique document ID for a book in Firestore
+  override fun getNewUid(): String {
+    return db.collection(collectionBooks).document().id
+  }
+  // Fetches the list of books from the Firestore collection
+  // If the task is successful, maps the Firestore documents to DataBook objects
+  // Calls OnSuccess with the list of books, or onFailure if the task fails
+  override fun getbook(OnSucess: (List<DataBook>) -> Unit, onFailure: (Exception) -> Unit) {
+    db.collection(collectionBooks).get().addOnCompleteListener { task ->
+      if (task.isSuccessful) {
+        // Maps Firestore documents to DataBook objects or returns an empty list
+        val books = task.result?.mapNotNull { document -> documenttoBooks(document) } ?: emptyList()
+        OnSucess(books)
+      } else {
+        task.exception?.let { e -> onFailure(e) }
+      }
+    }
+  }
+  // Adds a new book to the Firestore collection
+  // Calls OnSuccess if the operation is successful, otherwise onFailure with the exception
+  override fun addBooks(dataBook: DataBook, OnSucess: () -> Unit, onFailure: (Exception) -> Unit) {
+    performFirestoreOperation(
+        db.collection(collectionBooks).document(dataBook.Title).set(dataBook), OnSucess, onFailure)
+  }
+  // Updates an existing book in Firestore by replacing the document with the same title
+  // Uses performFirestoreOperation to handle success and failure
+  override fun updatebook(
+      dataBook: DataBook,
+      OnSucess: () -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    performFirestoreOperation(
+        db.collection(collectionBooks).document(dataBook.Title).set(dataBook), OnSucess, onFailure)
+  }
+  // Deletes a book from Firestore by its title
+  // Uses performFirestoreOperation to handle success and failure
+  override fun deletebooks(
+      id: String,
+      dataBook: DataBook,
+      OnSucess: () -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    performFirestoreOperation(
+        db.collection(collectionBooks).document(dataBook.ISBN).delete(), OnSucess, onFailure)
+  }
+  // Maps a Firestore document to a DataBook object
+  // If any required field is missing, returns null to avoid incomplete objects
+  fun documenttoBooks(document: DocumentSnapshot): DataBook? {
+    return try {
+      val title = document.getString("Title") ?: return null
+      val author = document.getString("Author") ?: return null
+      val description = document.getString("Description") ?: return null
+      val rating = document.getString("Rating") ?: return null
+      val photo = document.getString("Photo") ?: return null
+      val language = document.getString("Language") ?: return null
```

Comment:
Here the use of two language variables, one for the string and the other after being transformed to the type feels unnecessary, maybe mix the two to have a single variable, even maybe a small helper function that transforms it if necessary.